### ISSUE TITLE
Sphinx Documentation

### ISSUE THREAD

**Issue by jtanx** [fontforge/fontforge#4113]
So I'm basically going to bite the bullet and convert all the HTML docs under `doc` to Sphinx documentation ( -> convert to reStructuredText).

Raise any thoughts/concerns now, otherwise I'll just do it, assuming I haven't lost interest by the end.

**Comment by skef** [comment]
I definitely want to talk about this first. What is the reasoning and what does it buy the project?

**Comment by ctrlcctrlv** [comment]
My obvious question is "why not some flavor of Markdown?"

I've had some recent experience with reStructuredText and honestly consider it a pain in the ass compared to Markdown...the syntax is bizarre.

**Comment by jtanx** [comment]
* Not a fan of writing docs directly in html
* It means we can move on from the split between the old/new documentation that we currently have.
* Sphinx provides a richer experience, like you can search for stuff, you can cross reference stuff etc. etc.
* Markdown is ok, but you can't do some more complex things in it, so you end up writing html anyway

**Comment by ctrlcctrlv** [comment]
It is a problem to need to write HTML to do something complex?

**Comment by skef** [comment]
Comment: So I have a lot of experience with Pandoc Markdown. How about we go with that instead? 
Meta-comment: Are we just picking a future documentation format based on one preference? 
Side-comment: You "can't do some more complex things" with markdown. "Sphinx provides a richer experience." The latter isn't exactly denying that you can't do some more complex things with Sphinx.

**Comment by jtanx** [comment]
> It is a problem to need to write HTML to do something complex?

Depends on how often you have to fall back to HTML I guess. It's just not something I want to bother with if I'm going to do this for a bunch of existing docs.

> So I have a lot of experience with Pandoc Markdown. How about we go with that instead?

At this point I'd prefer rst because (correct me if I'm wrong), there's basically one variant. I don't want to deal with 'what flavour of markdown are we using today'. Sphinx supports multiple generation targets, not just html fwiw

Sphinx is designed for, well, writing documentation, which is why I'm using it. 

> Are we just picking a future documentation format based on one preference?

Someone has to (a) make a decision and (b) invest time to actually do it, so basically yes.

> You "can't do some more complex things" with markdown. "Sphinx provides a richer experience." The latter isn't exactly denying that you can't do some more complex things with Sphinx.

Hmm I'm not sure I quite understand this one. Anyway what I meant is Sphinx out of the box provides for things like easy cross-referencing, documentation of functions, code blocks with code highlighting, TOC generation etc. which I like.

**Comment by skef** [comment]
> Hmm I'm not sure I quite understand this one. Anyway what I meant is Sphinx out of the box provides for things like easy cross-referencing, documentation of functions, code blocks with code highlighting, TOC generation etc. which I like.

There are extended versions of Markdown, like Pandoc's, that have better overall functionality than the one chosen the last time. As you note Sphinx has better support for code documentation, but that's actually a small part of what the current docs do. 

> Someone has to (a) make a decision and (b) invest time to actually do it, so basically yes.

I haven't talked about this really, but I also have an abandoned branch with re-worked documentation, although my focus was on trying to form a new hierarchy and make an updated and more general intro document. Note that I am not and will not make a "my work would be lost!" argument. 

In that process, however, I came to a number of (personal) conclusions:

1. With a decent CSS layer the anti-HTML argument is sort of silly. Just decide on the structures you want to support and implement them with classed divs/spans as necessary.
2. We don't know how many people are using the documentation "locally". Given how screwed up the website documentation is, I wouldn't be surprised if the number is high. I don't think we should make a decision to remove local viewing without discussion. 
3. Probably most importantly: The current HTML documentation has an idiosyncratic structure. I doubt it can just be "translated" into a radically different format without changing that structure significantly. You may be offering to do this, but I don't think all of those decisions should be unilateral any more than I think significant changes to the CharView UI (just as an example) should be unilateral.
4. I consider the largely abandoned Markdown translations to count against this sort of approach. Let's grant that you're willing and able to do the work: then what? However much people may be annoyed at HTML, it's those documents that have actually been maintained over time.

**Comment by jtanx** [comment]
There's nothing stopping you running the generation tool to get HTML and bundling that as part of the distribution (and kind of in part why I chose this over readthedocs).

I'm open to discussion on the structure of the documentation, but I don't think it's a bad start to do the conversion, culling anything obviously irrelevant now, and then going from there, even if that means we don't immediately end up completely cutting over to the 'new' format from the get go.

**Comment by skef** [comment]
I think one should pick one's documentation system in large part based on whether it can support the needed structure, so I guess I disagree. 

I mean, are we imagining the equivalent of a book? Of interlinked "chapters"? Is there going to be hierarchy, and if so how many degrees of hierarchy? (Much of the current documentation has a two-level hierarchical structure.)

**Comment by ctrlcctrlv** [comment]
I want to say, that although I consider reStructuredText to be arcane/bizarre, it's not worse than AOLpress output, so I can't really oppose if @jtanx is willing to do the work.

**Comment by skef** [comment]
I'm less worried about the input syntax than the flexibility of the output. E.g. here are a few questions just raised by cross-referencing: Suppose a side-bar similar to DesignWithFontforge was desirable. Does the cross-referencing system expose enough info to support that, possibly with extensions/hand-coding? Does the link/target format support support multiple decompositions or does it tie you to a particular granularity of file?

**Comment by jtanx** [comment]
Have a look at https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html, I'm pretty sure you can do all of that, I believe you can even do things like link to the documentation of a specific function/method.

**Comment by skef** [comment]
That's part of the picture. If there's a "resolution" phase for generating interlinks you might have to compose the input files 1:1 relative to the desired output decomposition. 

You've got similar questions with sectioning. [This](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections) implies that reStructuredText "wants" to be a book. Can you trade headings off and generation time, so that in one decomposition you might see "3.4" (for a section, say) and in a different one see "4"?

I've worked quite a bit with Pandoc and know the answers to these questions for it, but I don't know much about reStructuredText (except that most things I'm aware were generated from it look like Python manuals).

**Comment by skef** [comment]
None of this is to say that generating from reStructuredText can't work; I'm saying that the right order of decision making is:

1. Decide what sorts of decompositon(s) and style(s) are required or desirable.
2. Do some tests on fake files to ensure those are possible with the system.
3. Start translating into that system.

**Comment by skef** [comment]
BTW although I sound grumpy about the initial particulars I would like to help out with the meat and bones of this project. Some of my rewriting would probably be easy to work in (choice of format is irrelevant to that possibility). 

I have a script that takes an SFD and a glyph name and generates an SVG with either an embedded or external style, which I used for the diagrams in this: https://rawcdn.githack.com/fontforge/fontforge/master/doc/html/stroke.html . It even has the little arrows on the first points. It wouldn't be hard to add an option to insert a cursor (like an arrow) at some coordinate. Also if you view the source of or inspect the dialog image you'll notice I also generated at 3 resolutions, which seems like a good convention for images that can't be SVGs. With this combination of features it's easier to make high quality diagrams that look good at any practical resolution.

**Comment by jtanx** [comment]
> You've got similar questions with sectioning. This implies that reStructuredText "wants" to be a book. Can you trade headings off and generation time, so that in one decomposition you might see "3.4" (for a section, say) and in a different one see "4"?

Not entirely sure you have that level of granularity, but it's not out of this realm if you *really* wanted to do that. If I understand you can write whatever extension you want to make sphinx to do something custom. At the least, I know it supports having section numbering turned on or off.

Anything you see done in Python documentation (which uses sphinx), you can do yourself.

I've only worked on (part of) the python api documentation, but this is what it looks like so far:

http://fontforgebuilds.sourceforge.net/docs/index.html
http://fontforgebuilds.sourceforge.net/docs/python/fontforge.html

If you want to know what the rst looks like, click on `Page source` in the bottom right hand corner.

In terms of what I've done so far, I'd probably split that single page used to describe the fontforge uber-module into one per class that exists in the module, just so it's more manageable. Probably also need to do something about how some functions have multiple signatures...

Outside of API documentation, well I guess it'd be interesting to see how it handles other stuff like that stroke documentation.

**Comment by jtanx** [comment]
>  If there's a "resolution" phase for generating interlinks you might have to compose the input files 1:1 relative to the desired output decomposition.

If you've named something, you can reference it from wherever without regard to where that other reference lives.

**Comment by skef** [comment]
@jtanx maybe I didn't understand the scope of this project. Are you thinking mostly, and perhaps exclusively, about the Python and native scripting pages? Because for just that use I don't have any particular objection to the Sphinx approach, and I don't see any particular reason the API docs have to be "stylistically aligned" with the other docs.

**Comment by jtanx** [comment]
I started with the scripting docs, but I don't see why it can't be adopted for everything else.

**Comment by jtanx** [comment]
The Python docs are more or less [complete](https://fontforgebuilds.sourceforge.io/docs/python/fontforge.html). I'm working off this branch: https://github.com/fontforge/fontforge/compare/master...jtanx:sphinx?expand=1

Let me know if you have any concerns/suggestions on this. Otherwise I'm going to continue with the rest...

**Comment by skef** [comment]
> Let me know if you have any concerns/suggestions on this. Otherwise I'm going to continue with the rest...

I'm not sure how or whether to respond to this. It seems like my going further down this road is likely to just be unproductively annoying. However, it also seems like not saying anything more is equivalent to approval, or at least wipes away what I've already said. 

Since the general case was unconvincing I'll start with a particular example. How do you expect to handle the (recurring) figure sequence semantic, as in https://fontforge.org/editspiro.html ? 

Because of its age this is currently implemented using a table, which is awkward and inflexible by today's standard but OK for the time. Today (based on my earlier research on this) you would use likely figure elements nested in a figure element, together with some well-thought-out CSS with an overall L-R block layout. https://www.sitepoint.com/quick-tip-the-right-way-to-use-figure-and-figcaption-elements/ The `figcaption` tag of the inner elements would carry the explanation. 

I haven't looked at Sphinx in enough detail to know whether it has a semantic for this; I'll just guess it probably doesn't given that the other similar frameworks I've worked with don't. I suppose one could try to build a large-element table, with all the contemporary problems that implies.

If that's right then the initial decisions would seem to imply the conclusion. HTML is awkward and best to avoid with a framework, the framework doesn't have the semantic, so for the remaining stuff: "beat it to fit, paint it to match". The last folks who went through this process picked a framework without a reasonable cross-reference or "navigation" system and produced stuff like this: https://fontforge.org/en-US/documentation/interface/ where when you go into the individual pages many of the links are broken and there's effective way "back out". A Sphinx-based project won't have that problem -- it's other stuff that will wind up flattened. Then at the end there will be a little CSS or configuration to choose whether the nav column goes down the left or the right and maybe put a logo somewhere. 

And everyone who first arrives the top-level docs page, especially the "designers scared off by FontForge's UI" that @eliheuer mentions [here](https://twitter.com/eliheuer/status/1222667574372192256), will immediately have the subliminal impression "this feels like an API manual", and they will be subliminally right. Many of these people could be eased into that UI, given the right sort of documentation. 

The "Design with FontForge" folks understood this problem and in my view did a very good job on the style and navigation level *with what they bit off*. The problem is that they didn't bite off quite enough, and produced something that isn't quite a complete introduction and that isn't really a bridge into the rest of the documentation. The style and structure are also limited by the scope -- it isn't flexible enough to just pour in the rest of the non-API documentation. So doing things right requires some CSS-level thought and development, including navigation strategies.

One very depressing aspect of this situation is that the two people arguing about it are among the small number of regular bug-fixers, while the work that is needed to get the docs into a close-to-ideal place could be done by contributors with a much less esoteric skill set. But I guess that's the state of the project right now.

**Comment by jtanx** [comment]
Thanks for pointing out an example.

> I haven't looked at Sphinx in enough detail to know whether it has a semantic for this; I'll just guess it probably doesn't given that the other similar frameworks I've worked with don't.

There is, in fact semantic support for [figures](https://docutils.sourceforge.io/docs/ref/rst/directives.html#figure). Granted, it's implemented using divs, but I don't think that's a big deal (see also sphinx-doc/sphinx#3526).

Anyway, like I said, if you really wanted to use a figure/figcaption element, there's nothing stopping you from writing an [extension](https://www.sphinx-doc.org/en/master/usage/extensions/index.html) (in Python) to do whatever you wanted. e.g. I heard they were parsing kerneldoc comments from the Linux kernel to generate Sphinx docs.

> The last folks who went through this process picked a framework without a reasonable cross-reference or "navigation" system and produced stuff like this: https://fontforge.org/en-US/documentation/interface/

If I had to guess this was more a quick 'n dirty effort to stuff the HTML through a markdown converter because that's what Jekyll supports. I don't think there's much to read into that to be honest.

Anyway, here's my go at that page: https://jtanx.github.io/ffdocs/editspiro.html https://github.com/jtanx/fontforge/blob/sphinx/doc/sphinx/editspiro.rst (with most links not working since I haven't done any of the other pages)

**Comment by skef** [comment]
@jtanx my point wasn't about the figure tag specifically, it was about the user-visible semantic of a figure sequence. Which your translation does flatten, so that when sequences are mixed in with other figures there will be no visual distinction.

**Comment by jtanx** [comment]
Hmm, I don't quite follow - can you give an example/diagram of where this becomes an issue?

**Comment by skef** [comment]
I guess the main one I'm aware of is on this page: https://fontforge.org/editexample3.html That's three groups of sequential figures with an additional figure or picture (I'm not sure how the "o" is classed).

**Comment by jtanx** [comment]
Right, well I guess it sounds like you're after a fluid grid.

**Comment by skef** [comment]
My past work in web design has been very ad hoc and I don't know much terminology. Searching around it seem like this would typically be considered a narrower problem than a fluid-grid design, which is usually used at the top level to accommodate different screen dimensions. 

There is a screen element here: you'd probably want to size the individual elements appropriately and then have them wrap as appropriate. Hence "L-R Block layout". It's not very hard for someone with reasonable experience to make CSS for this sort of case. However the present issue is which of the unusual semantic relations in the current docs will be retained.

**Comment by jtanx** [comment]
I'm pretty confident it's a well-solved problem (e.g. something like [bootstrap](https://getbootstrap.com/docs/4.0/layout/grid/) grids) or [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox).

How hard that is to do with sphinx, I don't know. But it's probably possible. Whether or not it's worth it, I don't know.

**Comment by skef** [comment]
But that more or less my point. Would you honestly say that it *wouldn't* be worth doing if Sphinx *did* support it out of the box? If not, then the question is raised by the choice of Sphinx, which is presumably driven by a set of things that Sphinx does do well out of the box. 

I'm sure there are frameworks for other purposes that don't include a code snippet semantic, and developers who work on the (presumably limited) APIs on projects using those frameworks have been asked how important it is, really, to line up each character of each line with one another.

This is how and why the result will wind up looking like an API manual: The system was designed by a set of people primarily dealing with the problem of API documentation and used by larger set of people who tend to see A B and C as important and D E and F as ultimately not that important.

**Comment by jtanx** [comment]
> Would you honestly say that it wouldn't be worth doing if Sphinx did support it out of the box? 

Well with the default theme, and at my resolution, it's narrow enough that there's not exactly a whole lot of place to place images side by side. You could of course run with another theme.

![image](https://user-images.githubusercontent.com/5137410/73592146-4888a580-454b-11ea-9a62-e298f494bb6f.png)

But in any case, it seems somewhat misguided to be discarding a framework solely based on this.

> This is how and why the result will wind up looking like an API manual:

I don't see this as being an issue.

The current proposals, if I have this right are:

* Keep the current HTML
* Use pandoc
* Use sphinx

Does pandoc provide styling? From what I saw it's basically just a conversion tool from one format to another.

In which case, for both keeping the existing HTML and using pandoc, you have to roll *everything* yourself.

I don't know about you, but in this day and age, I don't care for having to hand-craft my own css and layout from scratch. There are people and pre-existing solutions out there which would do the job far better than anything I could be bothered to invest the time into - in case it's not obvious, I'm not a front-end developer.

Sphinx is widely used, provides a decent looking set of themes, and comes with many features *out of the box* that make it easy to document something. All I have to do is provide it with some simple markup and it's done; I don't have to think hard about it at all. That's good enough for me,

**Comment by skef** [comment]
My proposal, for non-API-related content, is to decide on an design and feature set, implement that in CSS, and adapt the existing HTML to conform to it. This is where we disagree. This has already happened once, in a limited way, for Design with FontForge. 

In your favor is your offer to "do the work", but that *already happened*: we have the markdown copies. So really the prospect is:

1. You'll do the work with a somewhat better result, and
2. (More importantly) you'll delete the HTML versions, thus settling the issue.

**Comment by ctrlcctrlv** [comment]
Sorry @skef, I definitely cannot support adapting the existing HTML. When I told you Markdown was OK, that was with the expectation that HTML intrusions upon the Markdown would be rare.

I've come around on this issue and now back @jtanx. I did some more work on Kitty which required more exposure to reStructuredText. What once seemed arcane now seems sensible. He's willing to do all the work, I say let him&mdash;as he said, we can extend rST as needed, even with HTML

**Comment by skef** [comment]
The end point of this effort isn't going to be a bunch of rST extensions, it's going to be close-to-vanilla rST and a declaration of victory. 

Anyway, I've said my piece.

**Comment by ctrlcctrlv** [comment]
> close-to-vanilla rST and a declaration of victory.

I maintain pfaedit.org, so don't assume when I'll declare victory or not :-)

I have great respect for the existing documentation. In fact, my first exposure to this project's issue tracker was when @davelab6 changed the fontforge.org site, and I registered pfaedit.org in protest, which was George's domain whose registration lapsed. (Yes, I was immature in 2015.)

I'm going to expect fidelity. For example, it says right now "&copy; 2020 FontForge Developers". I think this should be changed to "&copy; 2020 George Williams, et al.", with "et al" linking to the AUTHORS file.

I was going to wait for the PR to pick these nits, but that's an example of what I'll look at.

**Comment by jtanx** [comment]
I'll rehash why I think keeping/adapting the html isn't a good idea.

* Web standards change all the time, and I don't think anyone here has the motivation to keep it reflective of that.
* rST/markdown read almost like plain text. It's easy to read and write.
* It's easy to screw up HTML syntax when you're hand rolling it. [This](https://github.com/jtanx/random/blob/master/ffrst/tools5.py) was the sort of crap I had to do to extract the required info, and even that only got me halfway, having  to pass it through a html linter first to deal with the broken html

Some things you also lose out on with hand-crafted html, such as the [search](https://jtanx.github.io/ffdocs/search.html) functionality and auto-generated [index](https://jtanx.github.io/ffdocs/genindex.html) you get for free.

Things like [i8n](http://www.sphinx-doc.org/en/stable/intl.html), you also get, so e.g. you could quite easily pass everything off to transifex or crowdin to make it easier to translate the documentation.

In other words, there's whole lot of benefits I'm seeing to using Sphinx and not a whole lot of drawbacks, vs what I get out of hand-crafted html. If the issue is really around how it looks, and we were really instead going to divest time in custom theming the existing html, then why not spend that effort to make a custom Sphinx theme instead?

**Comment by frank-trampe** [comment]
I'd normally say that, absent definitive arguments, we ought to go with whatever the person maintaining the component in question wants to use. But all three parties here are likely ongoing contributors to documentation, so that doesn't really help.

For what it's worth (and not much, since I don't do much on the documentation), I prefer writing markdown but share @skef's concerns that conversion might involve excessive manual effort or loss of some content or structure.

**Comment by skef** [comment]
I'm not going to work on this project. If the format is changed and code work calls for changing the docs in the future I'll edit whatever is current. And as I've indicated I don't want to discuss it further. 

The through-line of the introductory portions of the existing documentation is out of date, in that often focuses on issues that are no longer central (like PostScript fonts). It also, on my reading, doesn't really "cohere" -- for a novice there often isn't much context, or the subject changes in ways difficult to follow. I made an attempt at a restructuring in a branch, which I've rebased here: https://github.com/skef/fontforge/tree/docwork . It's a bit of a jumble, and it's all an unedited first draft, but if there is anything of use feel free to do so.

**Comment by jtanx** [comment]
For me, I see the scope of this change as a conversion of what's there, and less about trying to restructure the whole thing at the same time. If there's anything obviously irrelevant now I'll drop it, and if I see quick wins I'll probably do it, but I don't want to conflate the issue.

**Comment by jtanx** [comment]
Just an update on this. I've finished the bulk of the conversion, so content-wise it's more or less complete. I just need to fix up some references and clean up the structure a bit.

(I've updated https://jtanx.github.io/ffdocs/ - just beware of the above, I know that the structure's a bit of a mess at the moment).

**Comment by ctrlcctrlv** [comment]
Hey @jtanx. Sorry your comment got lost in the shuffle! I just want to say, I _**love**_ it! Very, very good work. Much better than editing AOLpress-generated files! You're even using LaTeX or something on the math page! When can we merge this? :)

**Comment by jtanx** [comment]
Yeah math is written in LaTeX, which is then rendered by mathjax. We could also optionally prerender them to images, but I thought the mathjax option was nicer (it does mean you have to have an internet connection though, unless we bundle the mathjax js library too).

I still need to make the structure a little bit more sane, then fixup some cross-page references. But it's close. With work commitments, maybe towards the weekend you'll see a pr.
